src/NuGet.Core/NuGet.Packaging/PackageExtraction/ZipArchiveExtensions.cs

```diff
@@ -85,5 +87,23 @@ public static void UpdateFileTimeFromEntry(this ZipArchiveEntry entry, string fi
                 }
             }
         }
+
+        public static void UpdateFilePermissionsFromEntry(this ZipArchiveEntry entry, string fileFullPath, ILogger logger)
+        {
+            if (RuntimeEnvironmentHelper.IsWindows)
+            {
+                return;
+            }
+            // Entry permissions are not restored to maintain backwards compatibility with .NET Core 1.x.
+            // (https://github.com/NuGet/Home/issues/4424)
+            // On .NET Core 1.x, all extracted files had default permissions of 766.
+            // The default on .NET Core 2.x has changed to 666.
+            // To avoid breaking executable files in existing packages (which don't have the x-bit set)
+            // we force the .NET Core 1.x default permissions.
+            chmod(fileFullPath, 0x1f6); // 0766
```

Comment:
Actually `.dll` .net libraries don't need the 'x' flag at all since the OS doesn't execute them.  What is needed is things like `apphost` which doesn't have an extension and `somescript.sh` which does have an extension.  Honestly I don't think we should be doing extension based checking, it would be very hard to get it right.

> Was this previously being set for everything?

Yes.  See https://github.com/dotnet/corefx/blob/release/1.1.0/src/System.IO.FileSystem/src/System/IO/UnixFileStream.cs#L116-L123

In 2.0, the code now lives here: https://github.com/dotnet/coreclr/blob/532c45a27aa2f804ed43448ef1666beeef6c96ae/src/mscorlib/shared/System/IO/FileStream.Unix.cs#L50-L57